core/lib/zksync_core/src/sync_layer/snapshots.rs

```diff
@@ -0,0 +1,285 @@
+            value: log.value,
+        })
+        .collect();
+    storage
+        .storage_logs_dal()
+        .append_storage_logs(miniblock_number, &[(H256::zero(), transformed_logs)])
+        .await;
+}
+
+async fn sync_tree_chunk(
+    recovery: &mut MerkleTreeRecovery<'_, &mut dyn PruneDatabase>,
+    storage_logs: &[SnapshotStorageLog],
+) {
+    let logs_for_merkle_tree = storage_logs
+        .iter()
+        .map(|log| RecoveryEntry {
+            key: log.key.hashed_key_u256(),
+            value: log.value,
+            leaf_index: log.enumeration_index,
+        })
+        .collect();
+
+    recovery.extend(logs_for_merkle_tree);
+}
+
+async fn sync_factory_deps_chunk(
+    storage: &mut StorageProcessor<'_>,
+    miniblock_number: MiniblockNumber,
+    factory_deps: Vec<SnapshotFactoryDependency>,
+) {
+    if !factory_deps.is_empty() {
+        let all_deps_hashmap: HashMap<H256, Vec<u8>> = factory_deps
+            .into_iter()
+            .map(|dep| (dep.bytecode_hash, dep.bytecode))
+            .collect();
+        storage
+            .storage_dal()
+            .insert_factory_deps(miniblock_number, &all_deps_hashmap)
+            .await;
+    }
+}
+
+pub async fn load_from_snapshot_if_needed(
+    storage: &mut StorageProcessor<'_>,
+    client: &dyn MainNodeClient,
+    merkle_tree_db_path: &String,
+) -> anyhow::Result<()> {
+    let mut applied_snapshot_status = storage
+        .applied_snapshot_status_dal()
+        .get_applied_snapshot_status()
+        .await
+        .unwrap();
+
+    if !storage.blocks_dal().is_genesis_needed().await.unwrap() && applied_snapshot_status.is_none()
+    {
+        tracing::info!("This node has already been initialized without a snapshot, skipping");
+        return Ok(());
+    }
+
+    if applied_snapshot_status.is_some() && applied_snapshot_status.as_ref().unwrap().is_finished {
+        tracing::info!("This node has already been initialized from a snapshot, skipping!");
+        return Ok(());
+    }
+
+    let snapshot_response = client.fetch_newest_snapshot().await.unwrap();
+    if snapshot_response.is_none() {
+        tracing::info!(
+            "Main node does not have any ready snapshots, skipping initialization from snapshot!"
+        );
+        return Ok(());
+    }
+    let snapshot = snapshot_response.unwrap();
+    let metadata = snapshot.metadata;
+    let l1_batch_number = metadata.l1_batch_number;
+    let miniblock_number = snapshot.miniblock_number;
+    tracing::info!("Found snapshot with data up to l1_batch {}, created at {}, storage_logs are divided into {} chunk(s)", l1_batch_number, metadata.generated_at, snapshot.chunks.len());
+
+    if applied_snapshot_status.is_none() {
+        applied_snapshot_status = Some(AppliedSnapshotStatus {
+            l1_batch_number: l1_batch_number,
+            is_finished: false,
+            last_finished_chunk_id: None,
+        });
+    };
+    let mut applied_snapshot_status = applied_snapshot_status.unwrap();
```

Comment:
do we somehow validate that `snapshot_response` is the same as it was when we failed?

Reply:
that is - main node could have created a n ewer snapshot